Pcap++/src/PfRingDeviceList.cpp

```diff
@@ -4,14 +4,53 @@
 
 #define LOG_MODULE PcapLogModulePfRingDevice
 
+#include <cstdio>
 #include "PfRingDeviceList.h"
 #include "SystemUtils.h"
+#include "DeviceUtils.h"
 #include "Logger.h"
 #include "pcap.h"
 #include "pfring.h"
 
 namespace pcpp
 {
+	/// @cond PCPP_INTERNAL
+
+	namespace
+	{
+		/**
+		 * @class PfRingCloseDeleter
+		 * A deleter that cleans up a pfring structure by calling pfring_close.
+		 */
+		struct PfRingCloseDeleter
+		{
+			void operator()(pfring* ptr) const { pfring_close(ptr); }
+		};
+
+		/**
+		 * Reads the ring version of a PF_RING handle.
+		 * @param[in] ring A PF_RING handle.
+		 * @return A string representation of the ring version.
+		 */
+		std::string readPfRingVersion(pfring* ring)
+		{
+			uint32_t version;
+			if (pfring_version(ring, &version) < 0)
+			{
+				throw std::runtime_error("Couldn't retrieve PF_RING version, pfring_version returned an error");
```

Comment:
I'm not sure it should throw an exception, it can just return an empty string

Reply:
true, probably "unknown"

Reply:
It can be "unknown" also, I don't have a strong opinion

Reply:
Eh, it can return an empty or "Unknown" string, since its internal function.

I preferred throwing an exception because if `pfring_version` fails, the version isn't really "read" is it?
In addition to that, how often do we expect the reading of the version to fail?

If we look at the only current call site at line 94, the "happy" path currently is:
1. Read string.
2. Save string to variable.
3. Carry on...

If we are returning an empty string, the path will always need to check if the string is empty or equal to a magic string ("Unknown") after step 2. With the exception, we aren't paying overhead if the read succeeds, only if it fails.

So the question is, are we expecting the read to fail like half the time, or once in a blue moon?
The usage of exceptions, essentially removes overhead from the happy path and adds **more** overhead to the exceptional path.
- If we are expecting the read to fail a lot, an error code approach might be better.
- If the read is expected to succeed most of the time an exception approach is better.

Reply:
I don't expect `pfring_version()` to fail after `pfring_open()` succeeded. Still, I think throwing an exception makes the code a bit more complex than it needs to be. The caller of `readPfRingVersion()` doesn't care what the version is - it could be any string, including an empty string or "Unknown". `readPfRingVersion()` can decide which string to return, and if `pfring_version()` fails it can write a message to the log

Reply:
The caller does care tho. Both before and after the refactor.

See the old lines 56-57. `calcPfRingVersion` is called for every device until the version string is populated, and on error it does not populate the version string. The refactor mirrors that behaviour by not writing the result of `readPfRingVersion`, if an exception occurs.

I guess, we can make the result be an empty string and write that regardless, as the `if` condition will still retry for the next device, but a populated error string will require another condition to ignore it.

Reply:
ah you're right, I missed that. So I think the method can return an empty string if it fails?

Reply:
sure, if we start using it at more places, then we might discuss exceptions again.

Reply:
Changed to empty string. dfa993d